crates/symbolicator/src/services/symcaches.rs

```diff
@@ -359,8 +368,164 @@ fn write_symcache(
 
-    let file = writer.into_inner().map_err(io::Error::from)?;
+    let mut file = writer.into_inner().map_err(io::Error::from)?;
+
+    if bcsymbolmap_handle.is_some() {
+        file.flush()?;
+        file.seek(SeekFrom::End(0))?;
+        file.write_all(SYMBOLMAP_MARKER)?;
+    }
     file.sync_all()?;
 
     Ok(())
 }
+
+#[cfg(test)]
+mod tests {
+    use std::fs;
+    use std::sync::Arc;
+
+    use symbolic::common::DebugId;
+    use uuid::Uuid;
+
+    use super::*;
+    use crate::cache::Caches;
+    use crate::config::{CacheConfigs, Config};
+    use crate::services::bitcode::BitcodeService;
+    use crate::services::DownloadService;
+    use crate::sources::{
+        CommonSourceConfig, DirectoryLayoutType, FilesystemSourceConfig, SourceConfig, SourceId,
+    };
+    use crate::test::{self, fixture};
+    use crate::utils::futures::ThreadPool;
+
+    /// Creates a `SymCacheActor` with the given cache directory
+    /// and timeout for download cache misses.
+    fn symcache_actor(cache_dir: PathBuf, timeout: Duration) -> SymCacheActor {
+        let mut cache_config = CacheConfigs::default();
+        cache_config.downloaded.retry_misses_after = Some(timeout);
+
+        let config = Arc::new(Config {
+            cache_dir: Some(cache_dir),
+            connect_to_reserved_ips: true,
+            caches: cache_config,
+            ..Default::default()
+        });
+
+        let cpu_pool = ThreadPool::new();
+        let caches = Caches::from_config(&config).unwrap();
+        caches.clear_tmp(&config).unwrap();
+        let downloader = DownloadService::new(config);
+        let objects = ObjectsActor::new(caches.object_meta, caches.objects, downloader.clone());
+        let bitcode = BitcodeService::new(caches.auxdifs, downloader);
+
+        SymCacheActor::new(caches.symcaches, objects, bitcode, cpu_pool)
+    }
+
+    /// Tests that a symcache is regenerated when it was created without a BcSymbolMap
+    /// and a BcSymbolMap has since become available.
+    ///
+    /// We construct a symcache 3 times under varying conditions:
+    /// 1. No symbol map is not there
+    /// 2. The symbol map is there, but its absence is still cached, so it is
+    ///    not downloaded
+    /// 3. The download cache has expired, so the symbol map is now
+    ///    actually available.
+    ///
+    /// Lookups in the symcache should return obfuscated names in
+    /// 1 and 2 and unobfuscated names in 3.
+    ///
+    /// 2 is specifically intended to make sure that the SymCacheActor
+    /// doesn't constantly try to download the symbol map.
+    ///
+    /// TODO: This test currently does not work because step 2
+    /// results in unobfuscated names. For some reason the
+    /// SymCacheActor downloads the symbol map immediately,
+    /// even though the negative cache should still be valid.
+    #[tokio::test]
+    async fn test_symcache_refresh() {
+        test::setup();
+
+        const TIMEOUT: Duration = Duration::from_millis(500);
```

Comment:
This constant is experimentally chosen so that the test works on my machine.

Reply:
CI seems happy.